File: src/dataStructures/stack/ts/stack.ts
import { StackProps } from "./types"
import { EmptyStackError, FullStackError } from "./utils"

export class Stack<Type> {
    private props: StackProps<Type>

    constructor(stackSize: number) {
        if (typeof stackSize !== "number") {
            throw new Error("The size of the stack must be a number")
        } else if (stackSize <= 0) {
            throw new Error("The size of the stack must be greater than 0")
        }

        this.props = {
            size: stackSize,
            values: []
        }
    }

    public push(value: Type): void {
        if (this.isFull()) {
            throw new FullStackError()
        }

        const stackPointer = this.props.values.length
        this.props.values[stackPointer] = value
    }

    public pop(): Type {
        if (this.isEmpty()) {
            throw new EmptyStackError()
        }

        const stackPointer = this.props.values.length - 1
        const poppedValue = this.props.values[stackPointer]
        this.props.values.length -= 1

        return poppedValue
    }

    public peek(): Type {
        if (this.isEmpty()) {
            throw new EmptyStackError()
        }
        return this.props.values[this.props.values.length - 1]
    }

    public isEmpty(): boolean {
        return this.props.values.length === 0
    }

    public isFull(): boolean {
        return this.props.values.length === this.props.size
    }

    public print(): void {
        if (this.isEmpty()) {
            throw new EmptyStackError()
        }

        let output = ""
        for (let i = this.props.values.length - 1; i >= 0; i--) {
            output += `[ ${JSON.stringify(this.props.values[i])} ]\n`
        }

        console.log(output)
    }
}

File: src/dataStructures/stack/c/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "stack.h"

struct Stack* createStack(int size){
    if (size <= 0){
        printf("Size must be greater than 0!\n");
        return NULL;
    }

    int *values = malloc(sizeof(int)*size);
    struct Stack *stack = malloc(sizeof(struct Stack));
    stack->size = size;
    stack->pointer = -1;
    stack->values = values;

    return stack;
}

int push(struct Stack *stack, int value){
    if (isFull(stack)){
        printf("Stack is full!\n");
        return -1;
    }

    stack->pointer++;
    stack->values[stack->pointer] = value;

    return 0;
}

int pop(struct Stack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    int value = stack->values[stack->pointer];
    stack->pointer--;

    return value;
}

int peek(struct Stack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    return stack->values[stack->pointer];
}

int print(struct Stack *stack){
    if (isEmpty(stack)){
        printf("Stack is empty!\n");
        return -1;
    }

    for (int i=stack->pointer; i>=0; i--){
        printf("[ %d ]\n", stack->values[i]);
    }
    printf("\n");

    return 0;
}

int isFull(struct Stack *stack){
    return stack->pointer == stack->size - 1;
}

int isEmpty(struct Stack *stack){
    return stack->pointer == -1;
}

